diff --git a/logic/src/main/java/pl/pwr/ludoSimulator/logic/Player.java b/logic/src/main/java/pl/pwr/ludoSimulator/logic/Player.java
index 7f7d406..11ef1f8 100644
--- a/logic/src/main/java/pl/pwr/ludoSimulator/logic/Player.java
@@ -21,4 +21,21 @@ public class Player  {
         this.startPosition = startPosition;
         this.endPosition = endPosition;
     }
-}
\ No newline at end of file
+  @Override
+  public int hashCode() {
+    return id;
+  }